src/parse.c

```diff
@@ -2697,6 +2697,22 @@ bool parse_chain(char *string, chain_t *chain)
 	return true;
 }
 
+
+bool parse_bindcode(char *name, xcb_keysym_t *keysym)
+{
+	for (unsigned int i = 0; i < LENGTH(nks_dict); i++) {
+		keysym_dict_t nks = nks_dict[i];
+		if(strstr(name,"0x")){
+				unsigned int intkey = (unsigned int)strtol(name, NULL, 0);
+				if(nks.keysym == intkey){
+					*keysym = nks.keysym;
+					return true;
+				}
+			}
+	}
+	return false;
+}
+	
```

Comment:
I don't see the point of this feature, but if we must merge this, let's at least use more sane code..

This will let you bind `super + 0x71poop` as `super + q`, not just `super + 0x71`.

It is also terribly inefficient: it will iterate the whole key dictionary for each in a HOTKEY that is not a modifier, even if name doesn't not even contain `0x`; checking if it contains `0x` at each iteration and converting it to a number at each iteration if it does. 

```suggestion
bool parse_bindcode(char *name, xcb_keysym_t *keysym)
{
	char *end;

	if (name[0] != '0' || name[1] != 'x') {
	        return false;
	}

	*keysym = (xcb_keysym_t)strtol(name, &end, 16);
	if (*keysym == 0 || *end != '\0') {
		return false;
	}

	for (unsigned int i = 0; i < LENGTH(nks_dict); i++) {
		if (nks_dict[i].keysym == *keysym) {
			return true;
		}
	}

	return false;
}

```

And let's rename the function to something else since i have never heard the term *bindcode*: these are just the numberic representations of keysyms in hexadecimal.

<details><summary>patch</summary><p>

```diff
diff --git a/src/parse.c b/src/parse.c
index 982ff45..687fa08 100644
--- a/src/parse.c
+++ b/src/parse.c
@@ -2697,22 +2697,28 @@ bool parse_chain(char *string, chain_t *chain)
 	return true;
 }
 
-
 bool parse_bindcode(char *name, xcb_keysym_t *keysym)
 {
+	char *end;
+
+	if (name[0] != '0' || name[1] != 'x') {
+		return false;
+	}
+
+	*keysym = (xcb_keysym_t)strtol(name, &end, 16);
+	if (*keysym == 0 || *end != '\0') {
+		return false;
+	}
+
 	for (unsigned int i = 0; i < LENGTH(nks_dict); i++) {
-		keysym_dict_t nks = nks_dict[i];
-		if(strstr(name,"0x")){
-				unsigned int intkey = (unsigned int)strtol(name, NULL, 0);
-				if(nks.keysym == intkey){
-					*keysym = nks.keysym;
-					return true;
-				}
-			}
+		if (nks_dict[i].keysym == *keysym) {
+			return true;
+		}
 	}
+
 	return false;
 }
-	
+
 bool parse_keysym(char *name, xcb_keysym_t *keysym)
 {
 	for (unsigned int i = 0; i < LENGTH(nks_dict); i++) {

```
</p></details>

Reply:
> I don't see the point of this feature: it doesn't seem very useful, but, if we want to add it, let's at least use more sane code..
> 
> This will let you bind `super + 0x71abcd` and `super + 113t0xabc` as `super + q`, not just `super + 0x71`.
> 
> It is also very inefficient: it will iterate the whole key dictionary for each word in a HOTKEY definition that is not identified as a modifier, even if name does not even contain `0x`; checking if it contains `0x` at each iteration and converting it to a number at each iteration if it does.
> 
> We should also rename the function to something else since i have never heard the term _bindcode_ and these are just numberic representations of keysyms in hexadecimal.
> 
> Cheers, emanuele6
> 
> patch

Sorry for closing this without any reason (though the code quality speaks for itself).
I wrote this ages ago when I had barely learnt C, let alone best practices. As such it isn't very good.
I'm still learning C properly at the moment but I'm not at the stage to fix this and for it to be merge-worthy.

Reply:
Some people mentioned that *bindcode* is `i3wm` lingo for binding a command to a keycode.

If that is what your were trying to achieve (probably since what this patch implements is quite unhelfpul), know that that is not what you implemented.

Keycodes are a number that identify a key (a physical key on the keyboard) independently of the keyboard layout.

What you implemented is simply an alternative way to bind to a keysym.

Take for example the "`q` key" and the `Super_L` key:

```
With the italian keyboard layout, if I press those keys, I get:
  keycode 24 (keysym 0x71, q)
  keycode 133 (keysym 0xffeb, Super_L)
With the frech keyboard layout, if I press the same keys, I get:
  keycode 24 (keysym 0x61, a)
  keycode 133 (keysym 0xffeb, Super_L)
```

With this patch, I am allowed to write:
```sh
alt + shift + q
	bspc node -c
alt + a
	urxvt
Super_L
	dmenu_run
```
as:
```sh
alt + shift + 0x71
	bspc node -c
alt + 0x61
	urxvt
0xffeb
	dmenu_run
```

These hexadecimal numbers are just the internal numeric representation of the keysyms for `q` and `Super_L` (in the french and italian keyboard layouts), not keycodes; the keycodes are 133 for `Super_L` and 24 for `q` (or `a` in the french keyboard layout). As it's also clear from the code, there is no difference between using the names or the hexadecimal numbers.

Using these numbers instead of the keysym names is not very helpful because it simply makes the configuration harder to read without even having the benefit of making the configuration keyboard layout independent.

So, if implementing "bind to keycode" was your goal, you will have to redo all from scratch.

Also, people that are complaining about not being able to bind directly to keycodes in #216 don't seem to realise that if they want their `sxhkd` configurations to be keyboard independant, they can just start their `sxhkd` **without** the `-m <count>` option and configure their hotkeys based on the US keyboard layout.
`sxhkd` will not change keycodes for their hotkeys binded to keysyms if they change keyboard layout at runtime; `sxhkd` only does that *<count>* times if the `-m <count>` option is passed, or every time if `-m -1` is passed; otherwise it will always assume the US keyboard layout.

So there is not even really a need to implement this "bind to keycode" feature that will surely be more complex and awkward to implement than this ~20 lines patch.

Cheers,
  emanuele6